test/cgroups.bats

```diff
@@ -10,19 +10,47 @@ function teardown() {
 	cleanup_test
 }
 
-@test "pids limit" {
-	if ! grep -qEw ^pids /proc/cgroups; then
-		skip "pids cgroup controller is not available"
-	fi
-	CONTAINER_PIDS_LIMIT=1234 start_crio
+function run_pids_limit_test() {
+	local set_limit="$1"
+	local expected_limit="$2"
 
+	CONTAINER_PIDS_LIMIT="$set_limit" start_crio
 	jq '	  .command'='["/bin/sleep", "600"]' \
 		"$TESTDATA"/container_config.json > "$TESTDIR"/container_pids_limit.json
 
 	ctr_id=$(crictl run "$TESTDIR"/container_pids_limit.json "$TESTDATA"/sandbox_config.json)
 
 	output=$(crictl exec --sync "$ctr_id" sh -c 'cat /sys/fs/cgroup/pids/pids.max 2>/dev/null || cat /sys/fs/cgroup/pids.max')
-	[[ "$output" == "1234" ]]
+	[[ "$output" == "$expected_limit" ]]
+}
+
+
+@test "pids limit" {
+	if ! grep -qEw ^pids /proc/cgroups; then
+		skip "pids cgroup controller is not available"
+	fi
+	run_pids_limit_test "1234" "1234"
+}
+
+@test "negative pids limit" {
+	if ! grep -qEw ^pids /proc/cgroups; then
+		skip "pids cgroup controller is not available"
+	fi
+	run_pids_limit_test "-1" "max"
+}
+
+@test "zero pids limit" {
+	if ! grep -qEw ^pids /proc/cgroups; then
+		skip "pids cgroup controller is not available"
+	fi
+	pids_root=/sys/fs/cgroup/pids
+	if is_cgroup_v2; then
+		pids_root=/sys/fs/cgroup
+	fi
+
+	# pids_limit of 0 sets the pids.max to that of pid 1
```

Comment:
This is not entirely true -- I think the limit is taken from the parent cgroup, not the one that init belongs to.

It's a coincidence that systemd sets the same limits for all cgroups.

Reply:
I just checked on Ubuntu 20.04, init is not put to a pids cgroup:

```
kir@ubu2004:~$ grep pids /proc/1/cgroup 
8:pids:/
kir@ubu2004:~$ ls -l /sys/fs/cgroup/pids/
total 0
-rw-r--r--  1 root root 0 Mar  3 19:06 cgroup.clone_children
-rw-r--r--  1 root root 0 Mar  3 19:04 cgroup.procs
-r--r--r--  1 root root 0 Mar  3 19:06 cgroup.sane_behavior
-rw-r--r--  1 root root 0 Mar  3 19:06 notify_on_release
-rw-r--r--  1 root root 0 Mar  3 19:06 release_agent
drwxr-xr-x 38 root root 0 Mar  3 19:04 system.slice
-rw-r--r--  1 root root 0 Mar  3 19:06 tasks
drwxr-xr-x  3 root root 0 Mar  3 19:04 user.slice
```

i.e. your method would not work.

I suggest using /proc/self/cgroup instead.

Reply:
/proc/self/cgroup fails when I run it locally though, as the container inherits it from systemd, but my user has "max" configured
```
 sudo bats test/cgroups.bats -t -f zero                                                                                                                                                                                                     1..1                                                                                                                                                                                                                                          not ok 1 zero pids limit                                                                                                                                                                                                                      ...                                                                
# 37725 max
```
(that came from me adding an echo before the test)

Reply:
```
not ok 4 zero pids limit
# (in test file ./cgroups.bats, line 55)
#   `parent_limit=$(cat "$pids_root$(awk -v lookfor="$lookfor" -F : '$2 == lookfor {print $3; exit}' /proc/1/cgroup)/pids.max")' failed
# cat: /sys/fs/cgroup/pids//pids.max: No such file or directory
```

As I said above, on Ubuntu 20.04, init is not put to a pids cgroup, and this is the reason why it's failing.

One way of figuring out the limit would be querying systemd. Here's an example from a couple of machines:

```console
kir@ubu2004:~$ systemctl show -p DefaultTasksMax
DefaultTasksMax=4592


[kir@kir-rhat cri-o]$ systemctl show -p DefaultTasksMax
DefaultTasksMax=18933
```

Reply:
oh yeah that seems like it's the way, updated